File: Data Structures & Algorithms/house-robber-ii/submission-1.java
class Solution {
    public int rob(int[] nums) {
        int n =nums.length;
        if(n==1)return nums[0];
        int[] dp = new int[n+1];
        dp[1] = nums[0];
        int ans = 0;
        for(int i=1;i<n-1;i++){
            dp[i+1] = Math.max(dp[i],dp[i-1]+nums[i]);
        }
        ans = Math.max(ans,dp[n-1]);
        dp = new int[n+1];
        for(int i=1;i<n;i++){
            dp[i+1] = Math.max(dp[i],dp[i-1]+nums[i]);
        }
        
        return Math.max(ans,dp[n]);

    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = static_cast<int>(nums.size());
        vector<int> left_product(n+1,1); 
        vector<int> right_product(n+1,1); 
        vector<int> ans(n,1); 
        for(int i=0;i<n;i++){
            left_product[i+1] = left_product[i]*nums[i];
        }
        for(int i=n-1;i>0;i--){
            right_product[i-1] = right_product[i]*nums[i];
            ans[i] = left_product[i]*right_product[i];
        }
        ans[0] = left_product[0]*right_product[0];
        return ans;
    }
};


